### DISCUSSION TITLE
Confused and curious

### DISCUSSION THREAD

**Discussion by sorohere** [TheAlgorithms/C-Plus-Plus#2515] [Q&A] [answered]
While using `<cassert>` header file, we can assert specific conditions in our code to catch potential bugs. However, what do you recommend for handling edge cases or scenarios that are not explicitly mentioned in the assertions? How would you ensure robustness and error handling in such situations?

**Comment by realstealthninja** [comment]
Could you explain one of such edge cases/scenarios further?

**Reply by sorohere** [reply]
what i am trying to say is, if a developer relies solely on the test cases that their code passes, they may overlook scenarios they haven't explicitly considered. These untested cases could potentially lead to failures or errors that their code isn't equipped to handle.

```
#include <iostream>

unsigned long long factorial(int n) {
    if (n < 0) {
        throw std::invalid_argument("Factorial is not defined for negative numbers");
    }
    if (n == 0 || n == 1) {
        return 1;
    }
    return n * factorial(n - 1);
}

int main() {
    // Test cases that the developer has tried
    std::cout << factorial(5) << std::endl;   // Expected output: 120
    std::cout << factorial(0) << std::endl;   // Expected output: 1

    // Missed test case
    std::cout << factorial(-3) << std::endl;  // Oops, this will throw an exception!
    
    return 0;
}
```

**Reply by realstealthninja** [reply, marked as answer]
It's the job of the reviewers to point out the missed test cases 🙂 if you spot something like that please do suggest a change by reviewing the pull request!

**Reply by sorohere** [reply]
okay, got it.
